### ISSUE TITLE
AsyncStorage and data saving: weird behaviour and crashes (Android)

### ISSUE THREAD

**Issue by ksegla** [facebook/react-native#10605]
Hi everyone

I have this weird problem where the saving of a variable containing 6000+ objects fails nicely (doesn't save but otherwise everything looks fine) while the same object with around 1000 objects just crashes the app. Any idea what could be causing this?

Also, more generally, I've spent the last 10 days (part time) trying to solve memory issues. Now I may be doing something wrong but there is definitely a need for better support for apps that do need to save tens of MB. **Everything was great till I tried to save some large objects.** I tried Realm but it is essentially worse: it crashes for large objects and takes up 10 times the space needed by AsyncStorage.
I tried react-native-fs but couldn't really make it work as it was trying to use folders that I didn't have. Also many people were complaining about the performance hit when it comes to reading large files. So it didn't seem worth the effort.

All in all, saving with AsyncStorage or using files is really shaky right now :(

PS: I'm building an app and I absolutely need to save some data that could go up to hundreds of MB. I expect most users to need around 20 MB or less though. Yes, I could go for a server side solution but I don't want my app to need a server and also sending such data would become expensive fast.

**Comment by hey99xx** [comment]
First thing is, please fill the template, is your issue iOS / Android, what React version etc.
You can find the template if you try to open another issue, and copy paste it here. Until then, you will likely get very slow responses since your question will not be tagged by anyone with authority.

Based on your explanation, this does sound like a real issue, but I'm not sure how much request will be to fix it. AsyncStorage is likely based on SQLite databases on devices instead of a file system, so you will most likely crash the database, or make it extremely slow on the next app starts. I would suggest going forward with either building your own file system native module or fix your issues with react-native-fs since AsyncStorage is not designed to handle megabytes of data in a blob.

**Comment by ksegla** [comment]
Thanks for the answer. It made me search for blob handling in React Native and I came up with react-native-fetch-blob which I'm going to test. I could probably spend some time and build a system native module for Android but it would be much harder for iOS (no experience with Objective C) and I do want to keep my project as cross platform as possible.

PS: I just saw the top part of the template and assumed it was just to redirect to productpains for feature request. my bad. Anyway I'm 0.34.1 and it's Android. Will update the OP later. Overall, I think there is much needed improvement with respect to blob handling and file saving and I hope later versions of React Native will address this at some point.

**Comment by brentvatne** [comment]
@ksegla - let us know what you come up with!

**Comment by ksegla** [comment]
@brentvatne 
I now know that whatever happens, I'll stick with file solutions. At least I can see what's in the file, its size etc. without having to hack some code. AsyncStorage issues are now pretty low in my list of things to fix about React Native. The issue is still there : relatively smaller objects crash the app while bigger ones are not saved but leave the app alone. Almost as if AsyncStorage tries hard and crashes for the first situation but just 'nopes out' of the second situation. 

Truth though is that I was indeed doing some things wrong. I sacrificed too much space for the sake of speed. I realised my mistake and redesigned my data structures to get my space complexity down from O(n2) to O(n). Huge difference. I no longer have real blobs (very rarely do I reach 1MB now). 

I settled on react-native-fetch-blob. It was able to save without problems up to 10MB of data but crashed for bigger stuff. The module has a streaming functionality that is supposed to accommodate big data but I couldn't make it work properly and it was no longer relevant after my data structure change. I also had some issues with getting the right directory (right now, I'm using a hardcoded path). The problem is not entirely solved but I'm confident it will work out. 

**tl;dr:** my data structure was a storage nightmare for the sake of speed. I fixed that and is using react-native-fetch-blob to save (now reasonably sized) files instead of relying on AsyncStorage

**Comment by hramos** [comment]
Hi there! This issue is being closed because it has been inactive for a while.

But don't worry, it will live on with Canny! Check out its new home: https://react-native.canny.io/feature-requests/p/asyncstorage-and-data-saving-weird-behaviour-and-crashes-android
